Add items, auth_services, end_services, and auth_ip_addresses

grepsedawk opened the pull request:
- Adds all models/controllers/routes for an authentication API
- Tosses in a few extras, like `better_errors` and `guard-rspec`
- Implements testing for Ruby `2.3.3` as well

resolves #2

grepsedawk committed:
Add items, auth_services, end_services, and auth_ip_addresses

Adds:
- models/test
- controllers
- api
- serializers
- request tests

Removes:
- unneeded assets/views

grepsedawk committed:
API Authentication now is functional

To generate an API key, go to console and type "ApiKey.generate!"
It will RETURN the ApiKey
To use, simply include api_key in your params

grepsedawk committed:
Install better_errors

grepsedawk committed:
Add guard-rspec

grepsedawk committed:
Add ruby 2.3.3 to Travis testing

Also, create the database

ekate commented:
@pachok here are my initial comments:
1.I saw you've open an issue to add mysql instead of sqlite. I would  expect you to do the change in that pull request.
2. In the database schema you've omitted the code column for auth_service and end_service. De-normalization might be a controversial issue so I didn't insist on it but I still want to see human generated 4 letter codes in the lookup tables. They will be used as query parameters. Also in ip_addresses table you decided to use de-normalization, e.g. column "name". In that particular case lookup table will make more sense. ip addresses are associated with different rooms and it is useful to be able to query by room and this query doesn't need to be fast
3. Why do you have api/v1 directory instead of putting everything api. That is internal application so we are not going to support versioning
4.As a minimum in the item controller "all" method should support pagination. Ideally it should be done for all controllers that list all records
5. I am not sure we need to use guard as usually it's not the part of our workflow, but I will let the whole group to decide what they think

ekate commented:
@pachok, I see you've just added mysql while I was writing a comment. Thank you

grepsedawk commented:
1. Sorry, I rebased master while you were working on the comment.
2.
    a. No problem, I can add the code column in a later migration. I didn't know how much we wanted to keep these human generated with it being a purely API based app.
    b. As far as the ip address, I actually switched the relationship to a more natural relationship. `auth_service` now ONLY relates to `auth_service`s (Shib, Facebook, Google, etc) and `auth_ip_address`es (172.34.235.21, 10.30.53.11, etc) relate directly to items.
3. I used `api/v1` because, if NYU is going to potentially have many different apps using this application, it makes sense to have the version namespace on the API. Even if we think we'll never need it, it's easier to keep the `v1` in the case NYU later wants to later support an older version of this API. The way I see it is it doesn't hurt to keep it there, but could hurt later. That being said, if we still want the `/v1` removed, let me know and I will remove it.
4. Added.
5. I use guard exclusively for rspec. If this is something we don't want, it can be removed pretty easily at any time without issue.

ekate commented:
2 yes, ips can be related directly to items but ip can be grouped by rooms, so I am asking you to add this lookup table
3 Please remove versioning

grepsedawk commented:
2. I just want to ensure I'm clear on what you're asking to do. Are you asking to do the comparable thing to IP's that I was asking (about auth_services) here: https://nyucomposers.slack.com/archives/general/p1486650386000021 ?
3. Wilco

grepsedawk committed:
Remove v1 from API per @ekate's request

grepsedawk committed:
Fix API error

grepsedawk committed:
Add IP Address Groups per @ekate's request

jgpawletko commented on app/controllers/api/api_controller.rb:
```diff
@@ -0,0 +1,17 @@
+    end
+
+    api_key_public = params[:api_key][0..23]
+    api_key_private = params[:api_key][24..48]
```
It looks like your total `api_key` length is 48 characters. Therefore, due to zero indexing, I believe this should be `params[:api_key][24..47]`.

grepsedawk replied:
Fixed, thanks!

jgpawletko commented on app/controllers/application_controller.rb:
```diff
@@ -10,4 +10,10 @@ def authentication_callback
   def authenticate!
     redirect_to '/auth/nyulibraries' unless session[:auth] && (session[:auth]['credentials']['expires'] == false || session[:auth]['credentials']['expires_at'] > Time.now.to_i)
```
Hi @pachonk ,
You might want to create a private helper method to encapsulate the logic after `unless`.
i.e., 
```
def valid_session?
  session[:auth] && 
    (session[:auth]['credentials']['expires'] == false || 
      session[:auth]['credentials']['expires_at'] > Time.now.to_i)
end
```
then your `authenticate!` method could be:
```
...
  redirect_to '/auth/nyulibraries' unless valid_session?
...
```

grepsedawk replied:
Yeah, I had definitely considered that, but the massive `unless` was temporary until we decided what would be considered "authenticated" to authz. In the next PR I intend to move that to a new function.

jgpawletko commented on db/migrate/20170217185047_create_api_keys.rb:
```diff
@@ -0,0 +1,9 @@
+  def change
+    create_table :api_keys do |t|
+      t.string :api_key_public
+      t.string :api_key_hash
```
is `:api_key_hash` the same as the private fragment of `params[:api_key]` or is there some transformation between the database attribute and the value used in the controller?

grepsedawk replied:
`api_key_hash` is (pseudo code) `hash(api_key_public + api_key_secret)`

jgpawletko commented on lib/ipv4_address_serializer.rb:
```diff
@@ -0,0 +1,9 @@
+  def self.dump(value)
+    IPAddr.new(value).to_i
+  end
+end
```
I know this is a picky comment, but it looks like you're missing a newline

grepsedawk replied:
Good catch, fixed.

jgpawletko commented on lib/ipv6_address_serializer.rb:
```diff
@@ -0,0 +1,9 @@
+  def self.dump(value)
+    IPAddr.new(value).hton
+  end
+end
```
I know this is a picky comment, but it looks like you're missing a newline

grepsedawk replied:
Not picky at all. Will fix!

jgpawletko commented on spec/controllers/api/api_controller_spec.rb:
```diff
@@ -0,0 +1,33 @@
+      end
+    end
+  end
+end
```
I know this is a picky comment, but it looks like you're missing a newline

jgpawletko reviewed (commented):
Hi Alex,
I'm still reviewing your code, but thought I'd pass along the comments I have so far.

jgpawletko commented:
Hi @pachonk ,
I don't want to hold this up anymore, but I do want to look at an ERD to make sure I understand how the schema hangs together.  

Also, do you see advantages to using `has_and_belongs_to_many` vs.  with `has_many :through` ?  I usually prefer `has_many :through` because it allows one to add additional attributes to associations if needed.

Thanks!

Hi @ekate, 
feel free to merge if you're OK with the PR.

Thanks!
